management/commands/aggregate_subject_trials.py

```diff
@@ -0,0 +1,526 @@
+        else:  # Old way of calculating the aggregate hash
+            # For sessions that have a trials table, add the task data files
+            raw_task_ds = Dataset.objects.filter(
+                session__in=sessions, default_dataset=True,
+                name__in=['_iblrig_taskSettings.raw.json', '_iblrig_taskData.raw.jsonable', '_ibl_experiment.description.yaml'])
+            # If we don't have task data for each session, that's a problem. This check is a little unnecessary
+            assert len(set(raw_task_ds.values_list('session', flat=True))) == len(sessions), 'not all sessions have raw task data'
+            # At minimum include trials table, which will differ if extraction code has changed
+            trials_ds = Dataset.objects.filter(
+                session__in=sessions, default_dataset=True, name='_ibl_trials.table.pqt')
+            # Compute the hash
+            hash_ds = (trials_ds | raw_task_ds).order_by('hash')
+            hash_str = ''.join(str(item) for pair in hash_ds.values_list('hash', 'id') for item in pair).encode('utf-8')
+
+            new_hash = hashlib.md5(hash_str).hexdigest()
+
+        return new_hash
+
+    def register_dataset(self, file_path, file_hash=None, aggregate_hash=None, file_size=None, user='root', revision=None):
+        """
+        Register an aggregate trials table dataset.
+
+        Parameters
+        ----------
+        file_path : pathlib.Path
+            An aggregate trials table file to register.
+        file_hash : str
+            The MD5 hash of the file.
+        aggregate_hash : str
+            An MD5 hash of dependent dataset hashes.
+        file_size : int
+            The file size in bytes.
+        user : str, User
+            The Alyx database user registering the dataset.
+        revision : str, Revision
+            The revision name to use for the dataset.
+
+        Returns
+        -------
+        Dataset
+            The Dataset record for the aggregate table.
+        pathlib.Path
+            The output file path.
+        """
+        if not file_path.exists():
+            raise FileNotFoundError(file_path)
+        assert all((self.output_path, self.subject)), 'subject and output path must be set'
+        # Get or create the dataset
+        dset, is_new = Dataset.objects.get_or_create(
+            name='_ibl_subjectTrials.table.pqt', collection='Subjects', default_dataset=True,
+            dataset_type=DatasetType.objects.get(name='subjectTrials.table'),
+            data_format=DataFormat.objects.get(name='parquet'), object_id=self.subject.id)
+
+        # Check if unchanged; whether new revision is required
+        if revision or not is_new:
+            unchanged = (
+                    file_hash and file_hash == dset.hash
+                    and aggregate_hash == (dset.json or {}).get('aggregate_hash')
+                    and file_size == dset.file_size)
+            if revision or (dset.is_protected and not unchanged):
```

Comment:
If the dataset is protected shouldn't we always make a new dataset regardless of whether the hashes all appear to be the same?

Reply:
If unchanged == True it means the dataset is guaranteed to be identical, if it's False it may or may not be identical. If the new dataset is literally identical I see no reason to make a new revision, unless a revision is explicitly passed.